### ISSUE TITLE
0519

### ISSUE THREAD

**Issue by yliuls** [MoonshotAI/MoBA#33]
import torch
import torch.nn.functional as F

# pad + window
def window_partition(x, shape, num_part=4):
    t, H, W = shape
    # 将输入reshape为四维 [t, H, W, d]
    x = x.view(t, H, W, -1)

    # 计算H和W方向的填充量
    pad_h = (num_part - H % num_part) % num_part
    pad_w = (num_part - W % num_part) % num_part

    # 在H和W的右侧进行填充
    pad = (0, 0,  # 不填充d维度
           0, pad_w,  # W维度右侧填充
           0, pad_h,  # H维度右侧填充
           0, 0)  # 不填充t维度
    x_pad = F.pad(x, pad)

    # 计算填充后的尺寸和窗口大小
    H_pad, W_pad = H + pad_h, W + pad_w
    window_h, window_w = H_pad // num_part, W_pad // num_part

    # 划分窗口并调整维度顺序
    x_window = x_pad.view(t, num_part, window_h, num_part, window_w, -1)
    x_window = x_window.permute(1, 3, 0, 2, 4, 5)  # [4, 4, t, window_h, window_w, d]
    x_window = x_window.reshape(-1, t, window_h, window_w, x_window.shape[-1])

    return x_window

# unpad
def window_unpad(x, original_H, original_W, local_rank, num_part=4):
    """
    Args:
        x: 输入窗口 [16, t, pad_h, pad_w, d]
        original_H: 原始高度（未填充）
        original_W: 原始宽度（未填充）
    Returns:
        out: 二维Tensor [S, d], S为各窗口有效长度之和
        lengths: 一维Tensor [16], 记录每个窗口的有效长度
    """
    # 计算填充后的尺寸和窗口大小
    pad_h = (num_part - original_H % num_part) % num_part
    pad_w = (num_part - original_W % num_part) % num_part
    H_pad = original_H + pad_h
    W_pad = original_W + pad_w
    window_h, window_w = H_pad // num_part, W_pad // num_part

    # 存储每个窗口处理后的数据和长度
    outputs = []
    actual_seq_len = []

    idx = 0
    # 遍历所有窗口（按原始划分顺序）
    for k in range(local_rank*2, local_rank*2+2):
        # 计算窗口在原始网格中的位置
        i = k // 4  # 行索引（H方向）
        j = k % 4   # 列索引（W方向）

        # 计算有效区域边界
        H_start = i * window_h
        H_end = min(H_start + window_h, original_H)
        valid_h = H_end - H_start

        W_start = j * window_w
        W_end = min(W_start + window_w, original_W)
        valid_w = W_end - W_start

        # 提取当前窗口并裁剪有效区域
        current_window = x[idx]  # [t, pad_h, pad_w, d]
        valid_window = current_window[:, :valid_h, :valid_w, :]  # [t, valid_h, valid_w, d]

        # 展平并记录
        flatten = valid_window.reshape(-1, valid_window.shape[-1])  # [t*valid_h*valid_w, d]
        outputs.append(flatten)
        actual_seq_len.append(flatten.shape[0])
        idx += 1

    # 拼接所有窗口数据
    out = torch.cat(outputs, dim=0)  # [S, d]
    actual_seq_len = torch.tensor(actual_seq_len, device=x.device)

    return out, actual_seq_len

# pad
def window_pad(x, original_H, original_W, actual_seq_len, local_rank, num_part=4):
    """
    将展平后的序列恢复为填充窗口结构, 使用lengths参数确保正确分割

    Args:
        x: 输入二维Tensor [S, d], S = sum(lengths)
        original_H: 原始高度(未填充)
        original_W: 原始宽度(未填充)
        lengths: 一维Tensor [16], 每个窗口的有效长度

    Returns:
        windows: 恢复后的窗口结构 [16, t, window_h, window_w, d]
    """
    device = x.device
    dtype = x.dtype

    # 计算填充后的尺寸
    H_pad = original_H + (num_part - original_H % num_part) % num_part
    W_pad = original_W + (num_part - original_W % num_part) % num_part
    window_h, window_w = H_pad // num_part, W_pad // num_part

    # 分割输入数据
    window_chunks = torch.split(x, actual_seq_len.tolist(), dim=0)

    # 存储重建后的窗口
    padded_windows = []

    idx = 0
    # 遍历所有窗口
    for k in (local_rank*2, local_rank*2+2):
        # 获取当前窗口的原始位置
        i = k // 4  # H方向索引
        j = k % 4   # W方向索引

        # 计算有效区域
        H_start = i * window_h
        H_end = min(H_start + window_h, original_H)
        valid_h = H_end - H_start

        W_start = j * window_w
        W_end = min(W_start + window_w, original_W)
        valid_w = W_end - W_start

        # 获取当前窗口的数据块
        chunk = window_chunks[idx]

        # 重塑为时空结构
        window_data = chunk.view(t, valid_h, valid_w, -1)

        # 创建填充容器
        padded = torch.zeros(
            (t, window_h, window_w, x.shape[-1]), device=device, dtype=dtype)

        # 填充有效区域
        padded[:, :valid_h, :valid_w, :] = window_data

        padded_windows.append(padded)

    return torch.stack(padded_windows, dim=0)

# reverse
def window_reverse(x, original_H, original_W, num_part=4):
    """
    Args:
        windows: tensor [16, t, window_h, window_w, d]
        original_H: 原始未填充的H尺寸
        original_W: 原始未填充的W尺寸
    Returns:
        out: 二维tensor [S, d], S = t*original_H*original_W
    """
    # 计算填充后的尺寸
    H_pad = original_H + (num_part - original_H % num_part) % num_part
    W_pad = original_W + (num_part - original_W % num_part) % num_part
    window_h, window_w = H_pad // num_part, W_pad // num_part

    # 恢复四维张量
    windows_reshaped = x.view(num_part, num_part, t, window_h, window_w, -1)
    x_pad = windows_reshaped.permute(2, 0, 3, 1, 4, 5).contiguous()
    x_pad = x_pad.view(t, H_pad, W_pad, -1)

    # 裁剪填充区域
    x = x_pad[:, :original_H, :original_W, :]

    # 转换为二维输出
    out = x.reshape(-1, x.shape[-1])

    return out

if __name__ == '__main__':
    t, H, W, d = 4, 46, 34, 1024
    rank = 0
    x = torch.randn(t*H*W, d)
    out1 = window_partition(x, shape=[t, H, W])
    out2, out3 = window_unpad(out1, H, W, rank)
    out4 = window_pad(out2, H, W, out3, rank)
    out5 = window_reverse(out4, H, W)
    print(out5.shape)
